### PR TITLE
Non-record: LoRA TTT exploration on SOTA base (negative result)

### PR DESCRIPTION
**Pull request by hmlizama**
## Summary

Explored combining LoRA Test-Time Training (TTT) with the current SOTA (thwu1 Int5-MLP + BigramHash + SWA, 1.1428 BPB). TTT adapts per-document during evaluation using rank-8 LoRA adapters on Q/V projections.

**Key finding: TTT does NOT improve SOTA.** Documented as informative negative result for the community.

## Experimental Results (1xH100)

| Method | val_bpb | Delta |
|--------|---------|-------|
| Standard eval | 1.1734 | baseline |
| TTT (chunk=256, seq=1024) | 1.3117 | +0.138 (worse) |
| TTT (chunk=256, seq=2048) | 1.3366 | +0.163 (worse) |

## Why TTT Fails on SOTA

1. **Context loss**: SOTA trains with seq_len=2048. TTT chunks into 256-token segments with 1024 window, losing long-range context
2. **Redundant local signals**: SmearGate + BigramHash already inject strong bigram context into embeddings
3. **Document boundary artifacts**: LoRA resets between documents, barely trains on short FineWeb docs

## Code Changes

Minimal modifications to train_gpt.py:
- Added LoRA support to CausalSelfAttention, Block, and GPT.forward
- Added BatchedLinearLoRA, BatchedTTTLoRA classes
- Added eval_val_ttt_lora evaluation function
- Standalone eval script (eval_ttt_only.py in experiments/)

## Takeaway

TTT is most effective on models with short context windows and weak local embeddings. On architectures that already maximize context (seq_len=2048) and inject strong bigram signals, TTT provides no additional benefit.

Built on PR #414 by @thwu1. TTT approach from @samacqua.

## Test plan
- [x] Smoke test 200 steps (1xH100)
- [x] 5108-step run with standard eval
- [x] 5108-step run with TTT eval (seq_len=1024)
- [x] 5108-step run with TTT eval (seq_len=2048)
- [ ] Full 8xH100 training run (pending compute credits)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: LoRA TTT exploration on SOTA base (negative result)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


**PR #658** — "Non-record: LoRA TTT exploration on SOTA base (negative result)"
**Head SHA:** 0d205f89284277b49898fabdc623d0a1672c4d26
**Author:** hmlizama
**Date:** 2026-03-24

---

## Check 1: N-gram family bug (target token in hash key) — CLEAN

`BigramHashEmbedding.bigram_hash` at line 614:

```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```

Hash key combines `t[i]` (current input token) XOR `t[i-1]` (previous input token). Both are input-side tokens from `x`; no target/label token (`y`) appears in the hash. This is standard BigramHash — explicitly listed as legal. CLEAN.

## Check 2: Pre-Quant TTT (multi-epoch AdamW on val_tokens without score-first) — CLEAN

No multi-epoch TTT during training. The `eval_val_ttt_lora` function runs only at final eval time (line 1439: `if args.ttt_enabled`), never during the training loop. The optimizer is `torch.optim.Adam` (not AdamW), single-pass, one gradient step per chunk. No pre-quantization multi-epoch adaptation detected. CLEAN.

## Check 3: Legal TTT (score-first-per-chunk) — CLEAN

The TTT loop (lines 1007–1028) follows the legal pattern:

**Verdict:** LOOKS CLEAN — legal TTT implementation.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by hmlizama** [comment]
Thanks for the thorough review @MatoTeziTanka! Glad the implementation checks out clean.
